### PR TITLE
Generalize input file representations of time sequences

### PR DESCRIPTION
**Pull request by wthrowe**
Everything that wants several times can now take either an explicit list or an infinite evenly spaced sequence.

@geoffrey4444 I think this should add what you need to try adaptive time stepping.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Time/TimeSequence.hpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  WRAPPED_PUPable_abstract(TimeSequence);  // NOLINT
+
+  using creatable_classes =
+      tmpl::list<TimeSequences::EvenlySpaced<T>, TimeSequences::Specified<T>>;
+
+  /// Returns the time in the sequence nearest to \p time, the time
+  /// before that, and the time after, in numerical order.  These
+  /// values allow a consumer to find the times in the sequence
+  /// bracketing the given time and give enough additional information
+  /// to that exact or roundoff matches can be handles however the
```
Comment (nilsdeppe) [inline src/Time/TimeSequence.hpp]: `handles` -> `handled`? 
`to` -> `so`?

File: src/Time/TimeSequence.hpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+///
+/// Holds all the TimeSequences
+namespace TimeSequences {
+template <typename T>
+class EvenlySpaced;
+template <typename T>
+class Specified;
+}  // namespace TimeSequences
+
+/// Represents a sequence of times.
```
Comment (nilsdeppe) [inline src/Time/TimeSequence.hpp:25]: Could you explain how the template parameter `T` is used in the documentation? E.g. do I choose it to be a `double`, `size_t`, `int`, `Time`, etc? And in what cases would I want to choose which type

File: src/Time/TimeSequence.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  ASSERT(std::is_signed_v<T> or current >= 0,
+         "Generated a negative value for an unsigned sequence.");
+
+  return {{std::is_signed_v<T> or previous >= 0 ? previous : std::optional<T>{},
+           current, next}};
+}
+
+template <typename T>
+void EvenlySpaced<T>::pup(PUP::er& p) noexcept {
+  p | interval_;
```
Comment (nilsdeppe) [inline src/Time/TimeSequence.cpp:49]: If I remember correctly you should call `TimeSequence<T>::pup(p);` first. https://charm.readthedocs.io/en/latest/charm++/manual.html#allocation-outside-pup

File: src/Time/TimeSequence.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  const auto next = closest_time + 1 == values_.end() ? std::optional<T>{}
+                                                      : *(closest_time + 1);
+  const auto previous = closest_time == values_.begin() ? std::optional<T>{}
+                                                        : *(closest_time - 1);
+  return {{previous, *closest_time, next}};
+}
+
+template <typename T>
+void Specified<T>::pup(PUP::er& p) noexcept {
+  p | values_;
```
Comment (nilsdeppe) [inline src/Time/TimeSequence.cpp:90]: Same comment about `TimeSequence<T>::pup(p)`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Couple minor things. This is really nice! Cleans up a lot of code :) Please go ahead and squash any changes you make

**Review by nilsdeppe — COMMENTED** [review]
fixup looks good. Please go ahead and squash :)

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**